### PR TITLE
Remove some collection decorators

### PR DESCRIPTION
**Pull request by victornoel**
This is for #1299 and #1304. 

Changes:
- I moved when possible the tests from collection to iterable tests
- I added some of the test that were removed in #1311 for #1303 to iterable's `Skipped` since it seemed a good idea

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#1315](https://github.com/yegor256/cactoos/issues/1315) is assigned to @fanifieiev[/z](https://www.0crat.com/u/fanifieiev), here is [why](https://www.0crat.com/footprint/C63314D6Z/d6bb12e0-4f33-489c-a436-ddbc070bf581); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: src/test/java/org/cactoos/iterable/SkippedTest.java [hunk 1]
```diff
@@ -23,31 +23,113 @@
...
     @Test
-    @SuppressWarnings("PMD.AvoidDuplicateLiterals")
     public void skipIterable() {
-        MatcherAssert.assertThat(
-            "Can't skip elements in iterable",
+        new Assertion<>(
+            "Must skip elements in iterable",
+            new Skipped<>(
+                2,
+                new IterableOf<>("one", "two", "three", "four")
```
Comment (fanifieiev) [inline src/test/java/org/cactoos/iterable/SkippedTest.java]: @victornoel Please extract these values into constants.

File: src/test/java/org/cactoos/iterable/SkippedTest.java [hunk 2]
```diff
@@ -23,31 +23,113 @@
...
+            ),
+            new IsEqual<>(
+                new IterableOf<>(
+                    "three",
```
Comment (fanifieiev) [inline src/test/java/org/cactoos/iterable/SkippedTest.java]: @victornoel Please extract these values into constants.

File: src/test/java/org/cactoos/iterable/SkippedTest.java [hunk 3]
```diff
@@ -23,31 +23,113 @@
...
+        ).affirm();
+    }
+
+    @Test
+    public void skipArray() {
+        new Assertion<>(
+            "Must skip elements in array",
+            new Skipped<>(
+                2,
+                "one", "two", "three", "four"
```
Comment (fanifieiev) [inline src/test/java/org/cactoos/iterable/SkippedTest.java]: @victornoel Please extract these values into constants.

File: src/test/java/org/cactoos/iterable/SkippedTest.java [hunk 4]
```diff
@@ -23,31 +23,113 @@
...
+            ),
+            new IsEqual<>(
+                new IterableOf<>(
+                    "three",
+                    "four"
+                )
+            )
+        ).affirm();
+    }
```
Comment (fanifieiev) [inline src/test/java/org/cactoos/iterable/SkippedTest.java]: @victornoel Please extract these values into constants.

File: src/test/java/org/cactoos/iterable/SkippedTest.java [hunk 5]
```diff
@@ -23,31 +23,113 @@
...
+
+    @Test
+    public void skipCollection() {
+        new Assertion<>(
+            "Must skip elements in collection",
+            new Skipped<>(
+                2,
+                new ListOf<>("one", "two", "three", "four")
```
Comment (fanifieiev) [inline src/test/java/org/cactoos/iterable/SkippedTest.java]: @victornoel Please extract these values into constants.

File: src/test/java/org/cactoos/iterable/SkippedTest.java [hunk 6]
```diff
@@ -23,31 +23,113 @@
...
+            ),
+            new IsEqual<>(
+                new IterableOf<>(
+                    "three",
```
Comment (fanifieiev) [inline src/test/java/org/cactoos/iterable/SkippedTest.java]: @victornoel Please extract these values into constants.

File: src/test/java/org/cactoos/iterable/SkippedTest.java [hunk 7]
```diff
@@ -23,31 +23,113 @@
...
+        ).affirm();
+    }
+
+    @Test
+    public void skippedAllElements() {
+        new Assertion<>(
+            "Must skip all elements",
             new Skipped<>(
                 2,
+                "one", "two"
```
Comment (fanifieiev) [inline src/test/java/org/cactoos/iterable/SkippedTest.java]: @victornoel Please extract these values into constants.

File: src/test/java/org/cactoos/iterable/SkippedTest.java [hunk 8]
```diff
@@ -23,31 +23,113 @@
...
+        ).affirm();
+    }
+
+    @Test
+    public void skippedMoreThanExists() {
+        new Assertion<>(
+            "Can't skip more than exists",
+            new Skipped<>(
+                Integer.MAX_VALUE,
+                "one", "two"
```
Comment (fanifieiev) [inline src/test/java/org/cactoos/iterable/SkippedTest.java]: @victornoel Please extract these values into constants.

File: src/test/java/org/cactoos/iterable/SkippedTest.java [hunk 9]
```diff
@@ -23,31 +23,113 @@
...
+    }
+
+    @Test
+    @SuppressWarnings("PMD.AvoidDuplicateLiterals")
+    public void skippedNegativeSize() {
+        new Assertion<>(
+            "Must process negative skipped size",
+            new Skipped<>(
+                -1,
                 "one", "two", "three", "four"
```
Comment (fanifieiev) [inline src/test/java/org/cactoos/iterable/SkippedTest.java]: @victornoel Please extract these values into constants.

File: src/test/java/org/cactoos/iterable/SkippedTest.java [hunk 10]
```diff
@@ -23,31 +23,113 @@
...
             ),
-            Matchers.contains(
-                "three",
-                "four"
+            new IsEqual<>(
+                new IterableOf<>(
+                    "one",
```
Comment (fanifieiev) [inline src/test/java/org/cactoos/iterable/SkippedTest.java]: @victornoel Please extract these values into constants.

File: src/test/java/org/cactoos/iterable/SkippedTest.java [hunk 11]
```diff
@@ -23,31 +23,113 @@
...
 import org.junit.Test;
+import org.llorllale.cactoos.matchers.Assertion;
 
 /**
  * Test Case for {@link Skipped}.
  *
  * @since 0.34
  * @checkstyle JavadocMethodCheck (500 lines)
  */
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
```
Comment (fanifieiev) [inline src/test/java/org/cactoos/iterable/SkippedTest.java]: @victornoel Please remove this one after extracting values into constants

**Comment by victornoel** [comment]
@fanifieiev using constants in tests is counter-productive, each test should be autonomous, so using constants:
- prevent that
- makes the code less readable
- does not provide anything except removing one suppress warning that is clearly not made for tests code

See https://www.yegor256.com/2016/05/03/test-methods-must-share-nothing.html for more details

**Comment by victornoel** [comment]
@paulodamaso please consider the comment above too, it is not the first time that I see such requests from REVs and I really believe it's going against the well being of the project to blindly try to follow all the warnings of PMD. Let's not mixup the spirit of the law with the letter of the law please…

**Comment by fanifieiev** [comment]
> @paulodamaso please consider the comment above too, it is not the first time that I see such requests from REVs and I really believe it's going against the well being of the project to blindly try to follow all the warnings of PMD. Let's not mixup the spirit of the law with the letter of the law please…

@victornoel @paulodamaso Well, we have to decide here and follow the same way everytime, because I recently had to fix similar issues raised by @paulodamaso .

**Comment by victornoel** [comment]
@paulodamaso ping?

**Comment by paulodamaso** [comment]
@victornoel I am in favor of constant usages for repetitive data, that's why:
- most of our test cases already share test data. Let me explain: we use the same values on almost all of our test cases (e.g. `0, 1, 2, "one", "two", "three"`) and these values are repeated. Why not put them into constants if they are present into most of our test cases if they are the same? Or else, you could alternatively do not use the same set of data for each test, avoiding the duplication.
- we can use context-sensitive naming for better test understanding: we have lots of criptic statuses and return codes which are not too obvious to the programmer (take HTTP statuses for example). It is much better to check a status against a constant with meaning (for example, `HTTP_OK`) instead some magic number (`200`). The first is much more readable than the latter.
- we have some repeatable data that can fit into constants, avoiding mistakes: `"text/plain"`, `"multipart/formdata"` and so on are examples of repeatable values that can be shared across tests and should be turned into constants.

Makes sense?

**Comment by fanifieiev** [comment]
@victornoel ping

**Comment by victornoel** [comment]
@paulodamaso I'm a bit bothered because I don't agree at all with what you presented, I think it doesn't make sense and it is useless work just for the appearance of quality while it will actually reduce the quality of the code.

I propose to either not do it, or add a `@todo` for removing the suppress warning and avoid tests inputs duplication by using different literals. I don't think it is a good idea to use literal constant for this. Here is why:

> most of our test cases already share test data. Let me explain: we use the same values on almost all of our test cases (e.g. 0, 1, 2, "one", "two", "three") and these values are repeated. Why not put them into constants if they are present into most of our test cases if they are the same? Or else, you could alternatively do not use the same set of data for each test, avoiding the duplication

Each tests rely on similar constants, but in terms of domain, they are just similar "by chance", not by design. This would be a mistake to intertwine all the tests because of that. And it will take much useles energy to avoid using the same terms (but it can be an acceptable solution though).

> we can use context-sensitive naming for better test understanding: we have lots of criptic statuses and return codes which are not too obvious to the programmer (take HTTP statuses for example). It is much better to check a status against a constant with meaning (for example, HTTP_OK) instead some magic number (200). The first is much more readable than the latter.
> we have some repeatable data that can fit into constants, avoiding mistakes: "text/plain", "multipart/formdata" and so on are examples of repeatable values that can be shared across tests and should be turned into constants.

I don't think this has anything to do with the constant used in the tests, in that case the duplication is duplication of the same domain concept, so it makes sense to factor it. FYI, even using literal constant for HTTP codes is not something that @yegor256 would recommend, see https://www.yegor256.com/2015/07/06/public-static-literals.html.

**Comment by paulodamaso** [comment]
> @paulodamaso I'm a bit bothered because I don't agree at all with what you presented, I think it doesn't make sense and it is useless work just for the appearance of quality while it will actually reduce the quality of the code.
> 
> I propose to either not do it, or add a `@todo` for removing the suppress warning and avoid tests inputs duplication by using different literals. I don't think it is a good idea to use literal constant for this. Here is why:
> 
> > most of our test cases already share test data. Let me explain: we use the same values on almost all of our test cases (e.g. 0, 1, 2, "one", "two", "three") and these values are repeated. Why not put them into constants if they are present into most of our test cases if they are the same? Or else, you could alternatively do not use the same set of data for each test, avoiding the duplication
> 
> Each tests rely on similar constants, but in terms of domain, they are just similar "by chance", not by design. This would be a mistake to intertwine all the tests because of that. And it will take much useles energy to avoid using the same terms (but it can be an acceptable solution though).

@victornoel I think that you understand perfectly what I meant: if these values are the same "by chance", why  not make them explicitly different then? I agree with you when you say that it would be a mistake to tie all tests, but my point is, using constants or not, when repeating the same we are already coupling those tests around these values. In fact, it just becomes clearer when we get to the conclusion that it is not a good idea to use a constant for them. So, pleae leave a puzzle so we can fix it later.
> 
> > we can use context-sensitive naming for better test understanding: we have lots of criptic statuses and return codes which are not too obvious to the programmer (take HTTP statuses for example). It is much better to check a status against a constant with meaning (for example, HTTP_OK) instead some magic number (200). The first is much more readable than the latter.
> > we have some repeatable data that can fit into constants, avoiding mistakes: "text/plain", "multipart/formdata" and so on are examples of repeatable values that can be shared across tests and should be turned into constants.
> 
> I don't think this has anything to do with the constant used in the tests, in that case the duplication is duplication of the same domain concept, so it makes sense to factor it. FYI, even using literal constant for HTTP codes is not something that @yegor256 would recommend, see https://www.yegor256.com/2015/07/06/public-static-literals.html.
@victornoel Yes, the optimal solution would be create classes for each of these members in cases which the values are the same in the domain context.

**Comment by victornoel** [comment]
@paulodamaso So if I understand well: it makes sense to have different literals in tests to avoid them being tied (by literal or by using constant, whatever) and have less strong tests?
This makes sense to me, I will add a todo in that direction :)

**Comment by victornoel** [comment]
@fanifieiev @paulodamaso thanks, it is done

**Comment by paulodamaso** [comment]
@rultor merge

**Comment by paulodamaso** [comment]
> @paulodamaso So if I understand well: it makes sense to have different literals in tests to avoid them being tied (by literal or by using constant, whatever) and have less strong tests?
> This makes sense to me, I will add a todo in that direction :)

@victornoel what do you mean by "less strong tests"?

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @fanifieiev[/z](https://www.0crat.com/u/fanifieiev), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
Code review was too long (16 days), architects (@paulodamaso) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

**Comment by sereshqua** [comment]
@0crat quality good

**Comment by victornoel** [comment]
> @victornoel what do you mean by "less strong tests"?

@paulodamaso I meant what I think you were meaning: that if we use different literals for each test, that would make the test more robust.

**Pull request merged**